decouple product view gallery images config from Gallery Block

davidverholen opened the pull request:
Last week we had the Problem, that we wanted to add some additional image types (picture tag for different viewports) to the Product View Gallery Images.

Since the creation of the Data array for the json configuration is hardcoded in the block, we had to implement a Plugin.
https://github.com/magento/magento2/blob/develop/app/code/Magento/Catalog/Block/Product/View/Gallery.php#L58

Although it also works with a Plugin, I thought it would be nicer if this configuration is decoupled from the Gallery Block, since this configuration also depends on the template having the images defined in the view.xml

Commit:
decouple product view gallery images config from Gallery Block

Commit:
Fix Product View Gallery Block Unit Tests

davidverholen commented:
a better approach would maybe be to set a flag in the view xml for "use as gallery image". Not sure about this. Are there any ideas yet?

another approach would be to have the "base" image ids defined in the framework and merge it with the used theme.

I think it's not good to rely on the view.xml definitions in core modules as you can't be sure that the image definition exists in the theme

vrann commented:
@davidverholen can you please merge this branch with the latest develop?

davidverholen commented:
@vrann just rebased it on the current develop

ishakhsuvarov commented on app/code/Magento/Catalog/Block/Product/View/Gallery.php:
```diff
@@ -28,19 +29,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\AbstractView
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
-        EncoderInterface $jsonEncoder,
+        \Magento\Framework\Json\EncoderInterface $jsonEncoder,
+        \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory,
```
Unfortunately, due to backwards compatibility policy, we can not accept changes which change constructor signature by introducing a new required parameter.
You may fix it like:
```php
use Magento\Framework\App\ObjectManager;
...
public function __construct(
    \New\Dependency $newDependency = null
) {
    $this->newDependency = $newDependency ?: ObjectManager::getInstance()->get(\New\Dependency::class)
}
```

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Gallery/ImagesConfigFactory.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * @var \Magento\Framework\Data\CollectionFactory
+     */
+    protected $_dataCollectionFactory;
```
We do not encourage inheritance-based API, please make this property private. Additionally, naming convention with underscores is considered legacy, please rename this property.

ishakhsuvarov commented on app/code/Magento/Catalog/Block/Product/View/Gallery.php:
```diff
@@ -28,19 +29,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\AbstractView
         \Magento\Catalog\Block\Product\Context $context,
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
-        EncoderInterface $jsonEncoder,
+        \Magento\Framework\Json\EncoderInterface $jsonEncoder,
```
We encourage use of `\Magento\Framework\Serialize\Serializer\Json` for json serialization.

ishakhsuvarov commented on app/code/Magento/ProductVideo/Block/Product/View/Gallery.php:
```diff
@@ -19,23 +19,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\Gallery
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
         \Magento\Framework\Json\EncoderInterface $jsonEncoder,
         \Magento\ProductVideo\Helper\Media $mediaHelper,
+        \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory,
```
Please inject with the `ObjectManager` as described above.

ishakhsuvarov commented:
@davidverholen Thanks for the PR, looks pretty good to me.
Please resolve the conflict, check review comments and Travic CI fails. If it's all good I am going to proceed with the merge asap, so that no new conflicts arise.
Thank you for your collaboration!

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=3401) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=3401) before we can accept your contribution.<br/>

Commit:
implement code review change requests

davidverholen commented:
@ishakhsuvarov thx for the review!

I just implemented most change requests, but I had to keep the Json Encoder I think since it's part of the original constructor.

ishakhsuvarov commented:
@davidverholen Thanks, we are going to continue with review and merge

ishakhsuvarov committed:
magento/magento2#3401: Merge branch 'develop' of github.com:magento/magento2 into feature/catalog-extensible-gallery-images

 Conflicts:
	app/code/Magento/Catalog/Block/Product/View/Gallery.php
	app/code/Magento/Catalog/etc/frontend/di.xml

ishakhsuvarov committed:
magento/magento2#3401: Decouple product view gallery images config from Gallery Block

 - Code style fixes

ishakhsuvarov commented:
@davidverholen 
Please sign the CLA for this PR, we can not proceed until it's signed.
Also, if you don't mind, please check the "Allow edits from maintainers" checkbox on this PR, so that I could push back some coding style fixes and develop merge I have made for your branch.

davidverholen commented:
@ishakhsuvarov  I tried to sign the CLA, not sure if it worked. It redirected me to the CLA page, I signed in and then it reloaded the page with the box in the middle beeing empty. After some time, I was redirected back to this pr but it's still "yellow" in the pr checks

ishakhsuvarov commented:
@davidverholen 
Looks like you have been working on it with someone else or with a different account. That may be the cause.

bragento commented:
@ishakhsuvarov good point, there is one commit I made with our company account

ishakhsuvarov committed:
magento/magento2#3401: decouple product view gallery images config from Gallery Block

 - Code Style improvements

ishakhsuvarov committed:
magento/magento2#3401: Merge branch 'develop' of github.com:magento/magento2 into feature/catalog-extensible-gallery-images

Ctucker9233 commented:
@magento-team Can this be backported to 2.2?
